rfcs/2021-12-06-10298-kubernetes_logs-rewrite

```diff
@@ -0,0 +1,113 @@
+as we want to retain the contents in the `Store` after receiving a DELETE event for
+the corresponding contents. This allows us to enrich events we receive from a Pod
+after it's been deleted but we still have an open file handle for it's logs.
+
+Some of our existing code in `src/kubernetes` is already taken directly from `kube`
+(an older version) without modification which lends itself to being replaced completely.
+
+## Rationale
+
+Rewriting the `kubernetes_logs` source to leverage the existing "official" Kubernetes
+client for Rust will give us a more stable and maintable foundation. We currently
+are unable to maintain and troubleshoot the `kubernetes_logs` source properly and
+leveraging `kube` directly will provide an existing community of users, tests, and
+experiences to improve our ability to support this feature in Vector.
+
+Relying on an external library will reduce the amount of code and complexity that
+currently exists within Vector and allow us to focus less on making to tools correct
+and more on ensuring we're using them properly.
+
+## Drawbacks
+
+Converting the underlying libraries to `kube` (where possible) doesn't guarantee
+a resolution for reported bugs, but it does shift a non-trivial amount of functionality
+out of our project and onto a specialized library.
+
+## Prior Art
+
+- [Vector `kubernetes_logs` source](https://github.com/vectordotdev/vector/tree/master/src/sources/kubernetes_logs)
+  - [Vector `kubernetes`](https://github.com/vectordotdev/vector/tree/master/src/kubernetes)
+- [Datadog Kubelet provider](https://github.com/DataDog/datadog-agent/blob/main/pkg/autodiscovery/providers/kubelet.go)
+- [Fluent-Bit Kubernetes filter](https://github.com/fluent/fluent-bit/tree/master/plugins/filter_kubernetes)
+
+## Alternatives
+
+### Not do this at all
+
+Troubleshooting and maintaining the existing source has been challenging, and there is
+a clear lack of understanding of the current implementation. The source is an important
+part of our OSS offering and frequently used with the Agent role. Not doing this impairs
+our ability to support Vector running as an Agent in Kubernetes and reflects poorly on
+our overall reputation.
+
+### Rewrite to integrate with the Kubelet API
+
+This avenue was recommended internally, but is a departure from our existing implementation,
+and thus contains more unknowns and increases the likelihood of breaking changes to the
+existing source.
+
+### Replace existing source with an enrichment transform
+
+This option is capable of being used regardless of what role Vector has been deployed
+in, but it is a larger change to implement and a larger update for users too. Longer
+term the source can be decoupled and introduced as a standalone enrichment component.
+
+## Outstanding Questions
+
+- Do we want to make this change with a version change of the source, opting in
+to the change with a `version` option in the configuration?
```

Comment:
If our tests pass, our soaks show no regression, and there are nu user-facing changes, then I see no reason why we'd want to 1) keep the old code around, and 2) have users be confused by this knob.

Or is there another reason you're thinking of why we might want to do this?

Reply:
Only as a more cautious style of transition, we could handle this internally by adding a new feature for the rewrite. I expect it's an overly cautious approach to expose it as a new "version" of the component 👍

Reply:
Yeah, I can see it as a derisking approach. I personally think the risk here is low enough that I don't think we need to keep the old code around and version it. This is in contrast to the disk buffer rewrite where it is risky enough that we want to keep both around for a little while at least.

Reply:
I'd add the answer to the open question here.